### DISCUSSION TITLE
Generic update

### DISCUSSION THREAD

**Discussion by aymanalqadhi** [diesel-rs/diesel#2969] [Q&A] [answered]
Hi everyone!
I was trying to implement a generic repository, and I have succeeded so far. However, I get an error when trying to implement a generic update function

My generic update code:
```rust
fn update(&self, item: T) -> Result<T>
where
    T: IntoUpdateTarget + AsChangeset<Target = T::Table>,
    Update<T, T>: LoadQuery<DbPooledConnection, T>,
{
    Ok(diesel::update(item.clone())
        .set(item)
        .get_result(&self.get_connection()?)?)
}
```

An example model:
```rust
#[derive(
    AsChangeset,
    Clone,
    Debug,
    Deserialize,
    Identifiable,
    Insertable,
    Queryable,
    Serialize,
)]
pub struct Service {
    #[serde(skip)]
    pub id: KeyType,
    pub title: String,
    pub description: String,
    pub available_begin: Option<chrono::NaiveTime>,
    pub available_end: Option<chrono::NaiveTime>,
}
```

The error I get when calling the update function:
``` rust
137 |   self.ctx.services().update(service)?;
    |                       ^^^^^^ the trait `diesel::associations::Identifiable` is not implemented for `Service`
```

**Comment by weiznich** [comment, marked as answer]
This error is expected as `#[derive(Identifiable)]` generates the following impl:

```rust
impl<'a> Identifiable for &'a Service {
 // …
}
```

As the corresponding [`IntoUpdateTarget`](https://docs.diesel.rs/1.4.x/diesel/query_builder/trait.IntoUpdateTarget.html#implementors) impl assumes that the type implements `Identifiable` this implies that the type must be a reference (or you must provide a manual impl of `Identifiable` for your type itself.) In concequence this means that the bound `T:  IntoUpdateTarget + AsChangeset<Target = T::Table>` is not fulfilled.

As your code example is extremely incomplete it's hard to exactly tell how to solve this. You can either try to add an manual impl of `Identifiable` to your `Service` struct or try to adjust that trait bound so that at least the `IntoUpdateTarget` part is bound to an reference.

**Reply by aymanalqadhi** [reply]
Thanks a lot. I solved the issue by providing a manual implementation of `Identifiable` to `Service`.
